### PR TITLE
Fix flaky TestExternalAuthenticator tests

### PR DESCRIPTION
**Pull request by s2lomon**
When testing against sharedToken(), we need to control all its usages in order to be able to assert the results in a proper way. When the tests were allowed to run concurrently, some of them were interfering with each other which made the token to be shared in an uncontrollable manner, which in turn has made tests failing in a fals negative way. 

The change switches off the concurrency between the tests and fixes a bit the way the ExternalAuthenticator is constructed, imitating how the authenticator is used on production.

### CODE REVIEW

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 1]
```diff
@@ -52,6 +52,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
```
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:52]: Please put pull request description as commit message.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 2]
```diff
@@ -187,16 +188,15 @@ public void testAuthenticationFromMultipleThreadsWithLocallyStoredToken()
     @Test(timeOut = 2000)
     public void testAuthenticationFromMultipleThreadsWithCachedToken()
     {
-        ExecutorService executor = newCachedThreadPool(daemonThreadsNamed(this.getClass().getName() + "%n"));
```
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:190]: Separate commit.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 3]
```diff
@@ -238,13 +239,12 @@ public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticat
     @Test(timeOut = 2000)
     public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticateTimesOut()
     {
-        MockRedirectHandler redirectHandler = new MockRedirectHandler()
-                .sleepOnRedirect(Duration.ofMillis(5));
```
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:242]: Why is this removed?

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 4]
```diff
@@ -238,13 +239,12 @@ public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticat
...
+        MockRedirectHandler redirectHandler = new MockRedirectHandler();
 
-        ExternalAuthenticator authenticator = new ExternalAuthenticator(redirectHandler, (uri, duration) -> TokenPollResult.pending(uri), KnownToken.memoryCached(), Duration.ofMillis(1));
```
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:244]: Does this refactor matter? If not I would revert.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java (hunk 1 above)
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:52]: Please add description to the commit message why it fixes the flaky test.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 5]
```diff
@@ -169,10 +169,10 @@ public void testAuthenticationFromMultipleThreadsWithLocallyStoredToken()
                 .withResult(URI.create("http://token.uri"), successful(new Token("valid-token-4")));
```
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:169]: Please explain in commit message what is motivation behind this refactor. That this follows actual usage pattern of `ExternalAuthenticator` is used by concurrent jdbc connections.

**Review by kokosing — APPROVED** [review]
% commit messages.

Please also annotate flaky tests.

**Comment by kokosing** [comment]
> Please also annotate flaky tests.

I mean the tests that failed now (not ones you fixed).

**Comment by kokosing** [comment]
@s2lomon It works. Please revert the last commit.
@lukasz-walkiewicz We are still waiting for your review.

**Comment by kokosing** [comment]
> It works.

Sorry I was wrong. Your changes to workflow.yaml are not correct and only CLA was checked. So we don't know if it fixes the flakiness. See:
image

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java (hunk 1 above)
Comment (lukasz-walkiewicz) [reply client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:52]: A couple of typos in the commit message.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 6]
```diff
@@ -239,18 +240,19 @@ public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticat
     public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticateTimesOut()
     {
         MockRedirectHandler redirectHandler = new MockRedirectHandler()
-                .sleepOnRedirect(Duration.ofMillis(5));
+                .sleepOnRedirect(Duration.ofSeconds(1));
 
         ExternalAuthenticator authenticator = new ExternalAuthenticator(redirectHandler, (uri, duration) -> TokenPollResult.pending(uri), KnownToken.memoryCached(), Duration.ofMillis(1));
```
Comment (lukasz-walkiewicz) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java]: Is this authenticator still being used?

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java (hunk 5 above)
Comment (lukasz-walkiewicz) [reply client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:169]: Typo in the commit message.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java (hunk 4 above)
Comment (lukasz-walkiewicz) [reply client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:244]: It looks like it belongs to the next commit.

**Review by lukasz-walkiewicz — APPROVED** [review]
It would be good to cleanup the commits history a bit but apart from that LGTM.

**Comment by kokosing** [comment]
I see that now it works: 
image

Please address last comments, and remove not needed commits and I would like to merge it.

File: client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java [hunk 7]
```diff
@@ -352,8 +354,14 @@ ThrowableAssert firstException()
 
         void assertThatNoExceptionsHasBeenThrown()
         {
-            assertThat(exceptions)
-                    .isEmpty();
+            try {
```
Comment (kokosing) [inline client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:357]: Please revert

Comment (s2lomon) [reply client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java:357]: It's important I feel, as without this we know nothing about the exceptions that has occurred during the test execution, so it's quite important to have this now, just to be prepared for the next round of flaky test, especially that anything that will come up right now, won't be easy to reproduce locally.
